### PR TITLE
Steve/pseudo ct

### PR DESCRIPTION
**Pull request by stevenkboyd**
@Besler I'd like to merge the pseudoCT code that I've completed, and I tried to add the proper tests. However, the test still fails and I can't see why. It's trying to read the dicom series that already exists in the Bonelab/tests/cli/data directory and create a temporary output directory, but I must be doing something wrong. Maybe it's obvious to you?

### CODE REVIEW

**Comment by Besler** [comment]
@stevenkboyd I pushed some code to your branch to fix this. Details below.

**Issue 1**
`ModuleNotFoundError: No module named 'tests.cli.config_cli'`
There was an issue with fetching the dicom directory in the code. Fixed.

**Issue 2**
`ModuleNotFoundError: No module named 'pydicom'`
Since we're adding a new Python dependancy here, so we should update `requirements.txt`.

With these fixes, I now get this error in your tests:
```bash
    0.15 Create the output DICOM directory.
    0.15 Output directory created.
    0.37 Found 46 slices.
    0.37 Reading in 46 slices.
    0.89 Creating a mask using 0 to 1 as range.
    0.91 Morphological closing function on mask (radius = 1).
    3.67 Calculating inverted log.
    3.68 Input image data range: -1000 to 3095 (4095.0).
    3.68 ERROR: Input data array cannot contain negative values.
```
Sounds like the input dicom image has negative values which you don't expect?

**Comment by stevenkboyd** [comment]
Hi Bryce,

Thanks for #1.

For #2, thanks for fixing that too. The error is because pseudoCT is designed to read in MRI data and write something similar to a CT scan (hence “pseudo” CT). Since MRI data is always positive, and the test DICOM data seems to be CT (not sure), it throws an error. I guess we could add an MRI dataset into the test data suite, but I thought it was easier to just allow it run on the CT data and throw an error. I welcome your thoughts if you think we should (or should not) add an MR dataset as a sample DICOM file.

Cheers,

Steve

--
Steven Boyd
PhD, Professor, University of Calgary
Director, McCaig Institute for Bone and Joint Health
Full Signature<http://people.ucalgary.ca/~skboyd/Steven_Boyd/Signature.html>

On Mar 2, 2020, at 10:02 AM, Bryce Besler <notifications@github.com<mailto:notifications@github.com>> wrote:


@stevenkboyd<https://github.com/stevenkboyd> I pushed some code to your branch to fix this. Details below.

Issue 1
ModuleNotFoundError: No module named 'tests.cli.config_cli'
There was an issue with fetching the dicom directory in the code. Fixed.

Issue 2
ModuleNotFoundError: No module named 'pydicom'
Since we're adding a new Python dependancy here, so we should update requirements.txt.

With these fixes, I now get this error in your tests:

    0.15 Create the output DICOM directory.
    0.15 Output directory created.
    0.37 Found 46 slices.
    0.37 Reading in 46 slices.
    0.89 Creating a mask using 0 to 1 as range.
    0.91 Morphological closing function on mask (radius = 1).
    3.67 Calculating inverted log.
    3.68 Input image data range: -1000 to 3095 (4095.0).
    3.68 ERROR: Input data array cannot contain negative values.

Sounds like the input dicom image has negative values which you don't expect?

**Comment by Besler** [comment]
Hi Steve,

Adding an MRI dataset makes sense to me. Maybe we have a knee image floating around?

Bryce

**Comment by Besler** [comment]
@stevenkboyd Thank you for the MRI image! It is now in the [BonelabData Repository](https://github.com/Bonelab/BonelabData/tree/master/data/COR_PD_3)

And, of course, a new error!
```
E
======================================================================
ERROR: Can run `blPseudoCT`
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/babesler/Projects/Builds/Bonelab/tests/cli/test_blPseudoCT.py", line 44, in setUp
    PseudoCT(**self.args)
  File "/Users/babesler/Projects/Builds/Bonelab/bonelab/cli/PseudoCT.py", line 233, in PseudoCT
    pydicom.dcmwrite(dcm_fn_output,ds,True)
  File "/Users/babesler/anaconda3/envs/bl/lib/python3.7/site-packages/pydicom/filewriter.py", line 992, in dcmwrite
    write_dataset(fp, get_item(dataset, slice(0x00010000, None)))
  File "/Users/babesler/anaconda3/envs/bl/lib/python3.7/site-packages/pydicom/filewriter.py", line 555, in write_dataset
    write_data_element(fp, dataset.get_item(tag), dataset_encoding)
  File "/Users/babesler/anaconda3/envs/bl/lib/python3.7/contextlib.py", line 130, in __exit__
    self.gen.throw(type, value, traceback)
  File "/Users/babesler/anaconda3/envs/bl/lib/python3.7/site-packages/pydicom/tag.py", line 37, in tag_in_exception
    raise type(ex)(msg)
ValueError: With tag (7fe0, 0010) got exception: (7FE0,0010) Pixel Data has an undefined length indicating that it's compressed, but the data isn't encapsulated as required. See pydicom.encaps.encapsulate() for more information
Traceback (most recent call last):
  File "/Users/babesler/anaconda3/envs/bl/lib/python3.7/site-packages/pydicom/tag.py", line 30, in tag_in_exception
    yield
  File "/Users/babesler/anaconda3/envs/bl/lib/python3.7/site-packages/pydicom/filewriter.py", line 555, in write_dataset
    write_data_element(fp, dataset.get_item(tag), dataset_encoding)
  File "/Users/babesler/anaconda3/envs/bl/lib/python3.7/site-packages/pydicom/filewriter.py", line 495, in write_data_element
    "(7FE0,0010) Pixel Data has an undefined length indicating "
ValueError: (7FE0,0010) Pixel Data has an undefined length indicating that it's compressed, but the data isn't encapsulated as required. See pydicom.encaps.encapsulate() for more information

-------------------- >> begin captured stdout << ---------------------
    0.04 Create the output DICOM directory.
    0.04 Output directory created.
    0.08 Found 26 slices.
    0.08 Reading in 26 slices.
    0.23 Creating a mask using 0 to 1 as range.
    0.24 Morphological closing function on mask (radius = 1).
    2.33 Calculating inverted log.
    2.33 Input image data range: 0 to 522 (522.0).
    2.43 Finished calculating inverted log.
    2.43 Apply mask.
    2.48 Writing output DICOM slices.

--------------------- >> end captured stdout << ----------------------

----------------------------------------------------------------------
Ran 1 test in 2.509s

FAILED (errors=1)
```

I resolved this error by searching the error and adding [this if statement](https://github.com/pydicom/pydicom/issues/738#issuecomment-421581350). Also needed to add `gdcm` as a dependancy.

File: bonelab/cli/PseudoCT.py [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+    window_width = (np.max(data_array)-np.min(data_array))*0.75
+    #message("Created new Study Instance UID:","{}".format(study_instance_uid))
+    #message("Created new Series Instance UID:","{}".format(series_instance_uid))
+    
+    PrintSample = 0
+    idx=0
+    for fname in filenames:
+      ds = pydicom.dcmread(fname,force=True)
+      if ds.file_meta.TransferSyntaxUID.is_compressed is True:
+        ds.decompress()
```
Comment (Besler) [inline bonelab/cli/PseudoCT.py]: Here is where I modified.

**Comment by stevenkboyd** [comment]
Great, thanks for finding and fixing that! :)
Steve
--
Steven Boyd
PhD, Professor, University of Calgary
Director, McCaig Institute for Bone and Joint Health
Full Signature<http://people.ucalgary.ca/~skboyd/Steven_Boyd/Signature.html>

On Mar 2, 2020, at 4:26 PM, Bryce Besler <notifications@github.com<mailto:notifications@github.com>> wrote:


@Besler commented on this pull request.

________________________________

In bonelab/cli/PseudoCT.py<https://github.com/Bonelab/Bonelab/pull/10#discussion_r386712848>:

> +      if ds.file_meta.TransferSyntaxUID.is_compressed is True:
+        ds.decompress()


Here is where I modified.

**Comment by Besler** [comment]
Had to rebase master to resolve merge conflicts. (Technical jargon for "make this branch go into master"). I'm going to verify the tests pass here and then feel free to merge, etc.

**Comment by Besler** [comment]
We're back in business! `gdcm` isn't distributed for Python 3.8 yet so I had to remove those tests. But, this works.

**Comment by stevenkboyd** [comment]
Excellent!!
--
Steven Boyd
PhD, Professor, University of Calgary
Director, McCaig Institute for Bone and Joint Health
Full Signature<http://people.ucalgary.ca/~skboyd/Steven_Boyd/Signature.html>


On Mar 2, 2020, at 9:07 PM, Bryce Besler <notifications@github.com<mailto:notifications@github.com>> wrote:


Merged #10<https://github.com/Bonelab/Bonelab/pull/10> into master.

**Pull request merged**